[FLINK-23864][docs] Add pulsar connector document (#17188)
apache/flink: docs/content/docs/connectors/datastream/pulsar.md

```java
// Start from the earliest available message in the topic.
StartCursor.earliest()

// Start from the latest available message in the topic
StartCursor.latest()

// Start from a specified message between the earliest and the latest.
// Pulsar connector would consume from the latest available message if the message id doesn't exist.
// The start message is included in consuming result. 
StartCursor.fromMessageId(MessageId)

// Include or exclude the start message by using the second boolean parameter.
StartCursor.fromMessageId(MessageId, boolean)

// Start from the specified message time (Message<byte[]>#getEventTime()).
StartCursor.fromMessageTime(long)
```

{{< hint info >}}
Each Pulsar message belongs to an ordered sequence on its topic.
The sequence ID (```MessageId```) of the message is ordered in that sequence.
MessageId contains some extra information on how the message was stored (the ledger, entry, partition),
you can create a MessageId by using ```DefaultImplementation.newMessageId(long ledgerId, long entryId, int partitionIndex)```
{{< /hint >}}

### Boundedness

Pulsar source is designed to support both streaming and batch running mode.

> By default, the PulsarSource is set to run in streaming manner, thus never stops until Flink job fails or is cancelled.
> You can use ```setBoundedStopCursor(StopCursor)``` to specify stopping position and set the source running in batch mode.
> When all partitions have reached their stopping position, the source will exit.

Reviewer: ```suggestion
By default, the ```PulsarSource``` is set to run in a streaming manner, thus it never stops until Flink job fails or is cancelled.
You can use ```setBoundedStopCursor(StopCursor)``` to specify stop position such that the source can run in batch mode.
When all partitions have reached their stopping position, the source will finish.
```

You can also set PulsarSource running in streaming mode, but still stop at the stopping position by using ```setUnboundedStopCursor(StopCursor)```.

> The source will exit when all partitions reach their specified stopping position.

Reviewer: ```suggestion
The source will finish when all partitions reach their specified stopping position.
```

```java
// Connector will never stop consuming
StopCursor.never()

// Stop at the latest available message in Pulsar when connector start consuming.
StopCursor.latest()

// Stop when connector meet a given message, or stop at a message which is produced after this given message.
StopCursor.atMessageId(MessageId)

// Stop but include the given message in consuming result.
StopCursor.afterMessageId(MessageId)

// Stop at the specified message time (Message<byte[]>#getEventTime()).
StopCursor.atEventTime(long)
```

### Configurable Options

In addition to config options described above, you can set arbitrary options for PulsarSource, PulsarClient,
PulsarAdmin and Pulsar Consumer by using ```setConfig(ConfigOption<T>, T)``` and ```setConfig(Configuration)```.

#### Pulsar Client Options

Pulsar connector use both [client API](https://pulsar.apache.org/docs/en/client-libraries-java/)
for creating the consuming instance. Pulsar connector extract most parts of Pulsar's ```ClientConfigurationData```,
which is required for creating a Pulsar client, as Flink config options in ```PulsarOptions```.

{{< generated/pulsar_client_configuration >}} 

#### Pulsar Admin Options

> [Admin API](https://pulsar.apache.org/docs/en/admin-api-overview/) is used for querying topic metadata
> and filter the desired topics when connector use topic pattern subscription. It would share most part of the

Reviewer: ```suggestion
[Admin API](https://pulsar.apache.org/docs/en/admin-api-overview/) is used for querying topic metadata
and for discovering the desired topics when connector use topic pattern subscription. It would share most part of the
```

config options with client API. The config options listed here is only used in admin API.
They are also defined in ```PulsarOptions```.

{{< generated/pulsar_admin_configuration >}}

#### Pulsar Consumer Options

> Pulsar provide Reader API and Consumer API for consuming messages in different scenarios.
> Pulsar connector use Consumer API. Pulsar connector extract most parts of Pulsar's ```ConsumerConfigurationData``` as Flink config options in ```PulsarSourceOptions```.

Reviewer: ```suggestion
In general, Pulsar provides Reader API and Consumer API for consuming messages in different scenarios.
Flink's Pulsar connector uses Consumer API. Pulsar connector extract most parts of Pulsar's ```ConsumerConfigurationData``` as Flink config options in ```PulsarSourceOptions```.
```

The partition discovery would be disabled in batch mode even if you set this option with a non-negative value.
{{< /hint >}}

### Event Time and Watermarks

By default, the message will use the timestamp embedded in Pulsar ```Message<byte[]>``` as the event time.
You can define your own ```WatermarkStrategy``` for extract event time from the message itself,
and emit watermark downstream:

```java
env.fromSource(pulsarSource, new CustomWatermarkStrategy(), "Pulsar Source With Custom Watermark Strategy")
```

[This documentation]({{< ref "docs/dev/datastream/event-time/generating_watermarks.md" >}}) describes
details about how to define a ```WatermarkStrategy```.

### Message Acknowledgement

When a subscription is created, Pulsar [retains](https://pulsar.apache.org/docs/en/concepts-architecture-overview/#persistent-storage) all messages, even if the consumer is disconnected.
The retained messages are discarded only when the connector acknowledges that all these messages are processed successfully.
Pulsar connector supports four types of subscription, this makes the acknowledge behaviors variety among different subscriptions.

#### Acknowledgement on Exclusive and Failover

> ```Exclusive``` and ```Failover``` supports cumulative acknowledgment, we only need to acknowledge
> the latest successfully consumed message. All the message before the given message would be marked
> with consumed status.

Reviewer: ```suggestion
```Exclusive``` and ```Failover``` supports cumulative acknowledgment: Flink only needs to acknowledge
the latest successfully consumed message. All the message before the given message would be marked
with consumed status.
```

> Pulsar source acknowledge the current consuming message when when checkpoints are **completed**,
> for ensuring the consistency between Flink's checkpoint state and committed position on Pulsar brokers.

Reviewer: ```suggestion
Pulsar source acknowledge the current consuming message when checkpoints are **completed**,
for ensuring the consistency between Flink's checkpoint state and committed position on Pulsar brokers.
```

If checkpointing is not enabled, Pulsar source would periodic acknowledge message.
It's configured by ```PulsarSourceOptions.PULSAR_AUTO_COMMIT_CURSOR_INTERVAL```.

Note that Pulsar source does **NOT** rely on committed positions for fault tolerance.
Acknowledge message is only for exposing the progress of consumer and monitoring in these two subscriptions.

#### Acknowledgement on Shared and Key_Shared

```Shared``` and ```Key_Shared``` need to acknowledge message one by one. We will acknowledge
message in a transaction and commit it to Pulsar.

You should enable transaction in Pulsar ```borker.conf``` file when using these two subscriptions in connector:

```text
transactionCoordinatorEnabled=true
```

> Pulsar transaction would be created with 3 hours as the timeout. Make sure the timeout is bigger than the
> checkpoint interval. You can change the transaction timeout by ```PulsarSourceOptions.PULSAR_TRANSACTION_TIMEOUT_MILLIS```

Reviewer: ```suggestion
Pulsar transaction would be created with 3 hours as the timeout. Make sure that ```timeout >>
checkpoint interval + maximum recovery time```. You can change the transaction timeout by ```PulsarSourceOptions.PULSAR_TRANSACTION_TIMEOUT_MILLIS```
```
